DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToHost/TopLevel/ModelCurveToSpeckleTopLevelConverter.cs

```diff
@@ -0,0 +1,110 @@
+using Objects;
+using Speckle.Converters.Common;
+using Speckle.Converters.Common.Objects;
+using Speckle.Converters.RevitShared.Helpers;
+using System.Collections;
+
+namespace Speckle.Converters.RevitShared.ToSpeckle;
+
+[NameAndRankValue(nameof(SOBR.Curve.ModelCurve), 0)]
+public class ModelCurveToHostTopLevelConverter : BaseTopLevelConverterToHost<SOBR.Curve.ModelCurve, DB.ModelCurve>
+{
+  private readonly ITypedConverter<ICurve, DB.CurveArray> _curveConverter;
+  private readonly IRevitConversionContextStack _contextStack;
+
+  public ModelCurveToHostTopLevelConverter(
+    ITypedConverter<ICurve, DB.CurveArray> curveConverter,
+    IRevitConversionContextStack conversionContext
+  )
+  {
+    _curveConverter = curveConverter;
+    _contextStack = conversionContext;
+  }
+
+  public override DB.ModelCurve Convert(SOBR.Curve.ModelCurve target) =>
+    ModelCurvesFromEnumerator(_curveConverter.Convert(target.baseCurve).GetEnumerator(), target.baseCurve);
+
+  private DB.ModelCurve ModelCurvesFromEnumerator(IEnumerator curveEnum, ICurve speckleLine)
+  {
+    DB.ModelCurve? revitCurve = null;
+    while (curveEnum.MoveNext() && curveEnum.Current != null)
+    {
+      var curve = (DB.Curve)curveEnum.Current;
+      // Curves must be bound in order to be valid model curves
+      if (!curve.IsBound)
+      {
+        curve.MakeBound(speckleLine.domain.start ?? 0, speckleLine.domain.end ?? Math.PI * 2);
+      }
+
+      if (_contextStack.Current.Document.IsFamilyDocument)
+      {
+        revitCurve = _contextStack.Current.Document.FamilyCreate.NewModelCurve(
+          curve,
+          NewSketchPlaneFromCurve(curve, _contextStack.Current.Document)
+        );
+      }
+      else
+      {
+        revitCurve = _contextStack.Current.Document.Create.NewModelCurve(
+          curve,
+          NewSketchPlaneFromCurve(curve, _contextStack.Current.Document)
+        );
+      }
+    }
+
+    return revitCurve.NotNull();
+  }
+
+  /// <summary>
+  /// Credits: Grevit
+  /// Creates a new Sketch Plane from a Curve
+  /// https://github.com/grevit-dev/Grevit/blob/3c7a5cc198e00dfa4cc1e892edba7c7afd1a3f84/Grevit.Revit/Utilities.cs#L402
+  /// </summary>
+  /// <param name="curve">Curve to get plane from</param>
+  /// <returns>Plane of the curve</returns>
+  private DB.SketchPlane NewSketchPlaneFromCurve(DB.Curve curve, DB.Document doc)
+  {
+    DB.XYZ startPoint = curve.GetEndPoint(0);
+    DB.XYZ endPoint = curve.GetEndPoint(1);
+
+    // If Start end Endpoint are the same check further points.
+    int i = 2;
+    while (startPoint == endPoint && endPoint != null)
+    {
+      endPoint = curve.GetEndPoint(i);
+      i++;
+    }
+
+    // Plane to return
+    DB.Plane plane;
+
+    // If Z Values are equal the Plane is XY
+    if (startPoint.Z == endPoint.NotNull().Z)
```

Comment:
only Z may be null?

Reply:
not the z value, the notNull check is on the entire point object

Reply:
Ahh of course and I guess it only needs doing once